indexers/keyvalue/LevelDBIndexer/tests/test_leveldbindexer.py

```diff
@@ -19,43 +20,169 @@ def rm_files(file_paths):
                 shutil.rmtree(file_path, ignore_errors=False, onerror=None)
 
 
-def run_test(indexer):
-    def create_document(doc_id, text, weight, length):
-        d = jina_pb2.Document()
-        d.id = doc_id
-        d.buffer = text.encode('utf8')
-        d.weight = weight
-        d.length = length
-        return d
-
-    with indexer as idx:
-        data = {
-            'd1': MessageToJson(create_document(1, 'cat', 0.1, 3)),
-            'd2': MessageToJson(create_document(2, 'dog', 0.2, 3)),
-            'd3': MessageToJson(create_document(3, 'bird', 0.3, 3)),
-        }
-        idx.add(data)
+def create_document(doc_id, text, weight, length):
+    d = Document()
+    d._document.id = (str(doc_id) * 16)[:16]
+    d.buffer = text.encode('utf8')
+    d.weight = weight
+    d.length = length
+    return d
+
+
+def get_documents(ids):
+    documents = [
+        [0, 'cat', 0.1, 3],
+        [1, 'dog', 0.2, 3],
+        [2, 'crow', 0.3, 4],
+        [3, 'pikachu', 0.4, 7],
+        [4, 'magikarp', 0.5, 8]
+    ]
+    id_to_document = {d[0]: d for d in documents}
+    data = [
+        (id_to_document[id][0], id_to_document[id], MessageToJson(create_document(*id_to_document[id])).encode('utf-8'))
+        for id in ids
+    ]
+    return zip(*data)
+
+
+def apply_action(idx, action):
+    method, key_to_id = action
+    with idx:
+        keys = key_to_id.keys()
+        ids = key_to_id.values()
+        ids, documents, documents_bytes = get_documents(ids)
+        if method == 'add':
+            idx.add(keys, documents_bytes)
+        elif method == 'delete':
+            idx.delete(keys)
+        elif method == 'update':
+            idx.update(keys, documents_bytes)
+        else:
+            print(f'method {method} is not implemented')
         idx.touch()
         idx.save()
-        save_abspath = idx.save_abspath
-        index_abspath = idx.index_abspath
-    assert os.path.exists(index_abspath)
-    assert os.path.exists(save_abspath)
 
+
+def apply_actions(save_abspath, index_abspath, actions):
+    for action in actions:
+        idx = BaseIndexer.load(save_abspath)
+        apply_action(idx, action)
+
+    return save_abspath, index_abspath
+
+
+def validate_positive_results(keys, documents, searcher):
```

Comment:
now that I see, these are not callbacks, these are just assertions, I think is easier to read them inside the `test` itself